### PR TITLE
feat: add longest and shortest side resize values

### LINKED ISSUE #357: [Feature]: shortest/longest Side mode & enlarge/Reduce only on resizing
### Description

Image

Just like this.

### Problem or Motivation

Batch enlarge/reduce (shortest/longest) side for certain x pixel. If some pics are already fit for the x pixel, rimage should not resize it.

### Expected behavior

_No response_

### Alternatives Considered

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by vlad-salone**
Closes #357.

Adds `100l` and `100s` to `--resize`, plus `--reduce-only` and `--enlarge-only` as aliases for the existing `--no-upscale` and `--no-downscale`. No new flags.

```sh
# longest side becomes 1000, aspect ratio preserved
rimage mozjpeg --resize 1000l ./images

# shortest side becomes 1000
rimage mozjpeg --resize 1000s ./images

# shrink oversized images only, anything already under 1000 keeps its original size
rimage mozjpeg --resize 1000l --reduce-only ./images
```

### Why the existing values are not enough

`100w` fixes the anchor to the width for every image in the batch. With mixed orientation that makes portrait images far taller than the landscape ones are wide, so nothing comes out at a consistent size. `100l` and `100s` pick the anchored side per image instead.

Same idea as ImageMagick's geometry suffixes: https://imagemagick.org/script/command-line-processing.php#geometry

### Implementation

Two new `ResizeValue` variants and one helper in `src/cli/preprocessors/resize/value.rs`. `map_dimensions` decides per image whether the target length lands on the width or the height, then derives the other side from the aspect ratio.

The skip conditions in `pipeline.rs` are untouched. They already do the right thing for these values, because both axes scale in the same direction, so `--no-upscale` skips an image whose relevant side is already at or below the target and `--no-downscale` is the mirror of that.

Three guards that the existing `100w` path does not have:

- The value has to be digits with the marker at one end, nothing else. Pulling the first run of digits out of anything else accepts typos like `1.5l`, which resizes down to a single pixel. `100w` still does this, tracked separately in #378.
- The derived side is clamped to a minimum of one pixel. Truncation puts it at zero on extremely elongated images, and `Resize` rejects a zero dimension.
- A side length of `0` is rejected at parse time rather than at resize time, so the error appears next to the input that caused it.

> **Note**: parsing lowercases the input first, so `1000L` and `1000l` are the same value. The check that rejects two anchors in one value now covers `w`, `h`, `l` and `s`, which replaces the old `w` plus `h` check and keeps `150wh` invalid.

### The --reduce-only and --enlarge-only aliases

`--no-upscale` and `--no-downscale` name the direction that is disabled, so reading either one means inverting twice to arrive at the mode that is left. The aliases name the mode directly, and match the wording batch converters use for the same pair of options, including the one in the screenshot on #357.

These are `visible_alias` entries, so the argument ids and all the flag handling stay exactly as they were. Nothing about the existing spellings changes.

```
--no-downscale    Disable downscaling when resizing. [aliases: --enlarge-only]
--no-upscale      Disable upscaling when resizing. [aliases: --reduce-only]
```

### Tests

34 new tests, all in the binary crate.

Unit tests in `value.rs` cover parsing (every spelling that `100w` accepts, case insensitivity, zero, missing length, conflicting anchors, and ten malformed values that must not parse), display round trips, and the dimension mapping across landscape, portrait and square inputs, including upscaling, the already fits no-op, the one pixel clamp, aspect ratio preservation over a grid of sizes, and degenerate images with a zero dimension.

Integration tests in `pipeline.rs` drive real CLI arguments through `operations()` and execute the queued operations, asserting both the resulting dimensions and whether a resize was queued at all. That second part is what pins the "already fits" behaviour, since leaving an image alone means no operation is queued.

The aliases get three tests: each one is asserted to produce results identical to the flag it aliases across several image sizes, and a third checks that each alias moves images in the direction its name promises, which is the one thing that could silently go wrong when naming a double negative.

Full suite, `cargo test --features build-binary`:

```
running 42 tests
test result: ok. 42 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 4.60s
running 56 tests
test result: ok. 56 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 16.47s
running 6 tests
test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.22s
running 5 tests
test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s
```

`cargo fmt` and `cargo clippy --features build-binary --all-targets` are clean.

### Manual check

Three fixtures, 800x400, 400x800 and 120x80:

```
--resize 200l:
  landscape.jpg  200x100
  portrait.jpg   100x200
  small.jpg      200x133

--resize 200s:
  landscape.jpg  400x200
  portrait.jpg   200x400
  small.jpg      300x200

--resize 200l --reduce-only:
  landscape.jpg  200x100
  portrait.jpg   100x200
  small.jpg      120x80

--resize 200l --enlarge-only:
  landscape.jpg  800x400
  portrait.jpg   400x800
  small.jpg      200x133
```

Rejected values:

```
1.5l       error: invalid value '1.5l' for '--resize <RESIZE>': Invalid resize value
200l300    error: invalid value '200l300' for '--resize <RESIZE>': Invalid resize value
200sfoo    error: invalid value '200sfoo' for '--resize <RESIZE>': Invalid resize value
0l         error: invalid value '0l' for '--resize <RESIZE>': Side length should be greater than 0
100ls      error: invalid value '100ls' for '--resize <RESIZE>': Resize value can be anchored to only one of width, height, longest or shortest side
```

---

@Mikachu2333 you raised the argument surface concern on the issue, and this is my attempt at the version that does not grow it. Is `1000l` / `1000s` fine as naming, or would you prefer something closer to the ImageMagick geometry syntax?

### CODE REVIEW

**Comment by Mikachu2333** [comment]
Very good.

**Comment by Mikachu2333** [comment]
May need to improve. 

```
Example:

A: 200*100
B: 400*2000
C: 800*900
```

The user want to limit the width(height) of the pic smaller than 400px, no matter how long the hight it might be.

```
After Example:

A_CHANGED: 200*100
B_CHANGED: 400*2000
C_CHANGED: 400*300
```

**Comment by vlad-salone** [comment]
> The user want to limit the width(height) of the pic smaller than 400px, no
> matter how long the hight it might be.

That case already works without a side value, `400w` with `--reduce` does exactly it:

```
input:                  after:
  A.jpg    200x100        A.jpg    200x100
  B.jpg    400x2000       B.jpg    400x2000
  C.jpg    800x900        C.jpg    400x450
```

I added your A/B/C case as a test in `pipeline.rs` so the behaviour stays pinned.

Does that cover what you meant, or were you after something the width anchor
cannot express?

**Comment by Mikachu2333** [comment]
> That case already works without a side value, `400w` with `--reduce` does exactly it:

good for me.

File: src/cli/preprocessors/resize/value.rs [hunk 1]
```diff
@@ -42,10 +50,63 @@ impl ResizeValue {
...
+///
+/// The marker may sit on either end of the digits, the same spellings the width
+/// and height values accept, but nothing else may appear in the value. Pulling
+/// the first run of digits out of anything else would silently accept typos
+/// like `1.5l`, which would resize down to a single pixel.
+///
+/// A zero length is rejected here rather than at resize time, because it can
+/// never produce a valid image and the error is clearer next to the input.
+fn parse_side(s: &str, marker: char) -> Result<usize, anyhow::Error> {
+    let Some(digits) = s.strip_prefix(marker).or_else(|| s.strip_suffix(marker)) else {
```
Comment (Mikachu2333) [inline src/cli/preprocessors/resize/value.rs:94]: Consider tightening `w/h` parsing also to sync with strictness of the new `l/s` values: `200wfoo` currently parses as `200w` because the branch takes the first run of digits, while `200lfoo` is rejected.
Rejecting trailing garbage for `w/h` too would keep error handling consistent.
This is just a suggestion which could be added in another PR.

Comment (vlad-salone) [reply src/cli/preprocessors/resize/value.rs:94]: Agreed, and it is already filed as #378 with the same reasoning.

I kept `w` and `h` out of this PR because tightening them changes released behavior rather than new syntax. `parse_side` takes the marker as an argument, so routing all four markers through it is most of the work.

One thing to decide there: `200wfoo` and `1.5w` clearly should fail, but `w=100` also stops parsing, and that works today. The documented spellings (`100w`, `w100`, `100 w`) are unaffected either way.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/cli/preprocessors/mod.rs [hunk 2]
```diff
@@ -49,10 +55,12 @@ impl Preprocessors for Command {
                     .overrides_with("no-downscale"),
                 #[cfg(feature = "resize")]
                 arg!(--"no-downscale" "Disable downscaling when resizing.")
-                    .long_help(indoc! {r#"Disable downscaling when resizing.
+                    .long_help(indoc! {r#"Disable downscaling when resizing. [aliases: --enlarge-only]
```
Comment (Mikachu2333) [inline src/cli/preprocessors/mod.rs:58]: Please add a test for passing `--no-upscale` together with `--no-downscale`. We have no test or docs about what would happen on this circumstance.
The clap overrides_with wiring makes the last flag win, and `operations()` then skips every resize, but nothing pins this.

File: src/cli/pipeline.rs [hunk 3]
```diff
@@ -456,3 +456,248 @@ pub fn encoder(name: &str, matches: &ArgMatches) -> Result<AvailableEncoders, Im
...
+        let queued = ops.values().filter(|op| op.name() == "fast resize").count();
+
+        for op in ops.values() {
+            op.execute(&mut img).unwrap();
+        }
+
+        (queued, img.dimensions())
+    }
+
+    #[test]
```
Comment (Mikachu2333) [inline src/cli/pipeline.rs:509]: Please add a test covering chained `--resize` values (e.g. `--resize 1000l --resize 500s`). Every target size is currently computed from the **original** image dimensions, so once the first resize changes the aspect ratio, the second target is derived from a stale ratio.
I wonder if it is a bug or intended feat.

**Review by Mikachu2333 — CHANGES_REQUESTED** [review]

File: src/cli/preprocessors/mod.rs (hunk 2 above)
Comment (vlad-salone) [reply src/cli/preprocessors/mod.rs:58]: Added in `disabling_both_directions_skips_every_resize`.

One correction on the mechanism: `overrides_with` does not come into it, so there is no last flag winning. `--downscale` overrides `--no-downscale` and `--upscale` overrides `--no-upscale`, but `--no-upscale` and `--no-downscale` are independent of each other. Both stay set, both directions are disabled, and the first skip condition drops every resize:

```rust
if (!downscale && !upscale) || (w == size.0 && h == size.1) {
```

Order makes no difference, which the test covers both ways round plus the `--reduce-only --enlarge-only` spelling:

```
rimage mozjpeg --resize 1000l --no-upscale --no-downscale ./f1t.jpg   -> 48x80 (unchanged)
rimage mozjpeg --resize 1000l --no-downscale --no-upscale ./f1t.jpg   -> 48x80 (unchanged)
```

Worth noting the combination silently does nothing rather than erroring. Making it a hard error seems better to me, but that is existing behavior so I left it alone here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/cli/pipeline.rs (hunk 3 above)
Comment (vlad-salone) [reply src/cli/pipeline.rs:509]: Added in `chained_resize_values_all_map_the_original_size`.

On bug or feature, I think bug. `size` is read once before the loop and every value maps it, so it is worse than a stale ratio: for a chain of pure resizes the last value fully overwrites the earlier ones, and the work they did is thrown away.

Source is 800x400:

```
--resize 100x400 --resize 200s   -> 400x200
--resize 1000l --resize 50%      -> 400x200
--resize 1000l --resize 500s     -> 1000x500
```

The first one is the clearest. `200s` maps 800x400 and gets 400x200, when the image it actually runs on is 100x400, where the shortest side is the width and the result would be 200x800. The second shows `50%` is a half of the source, not of the 1000x500 the previous value produced.

The third looks right, but only by luck: both values map the same source and happen to land on 1000x500.

Your example was picked from the one case that hides it. The tests cover all three so the difference is visible.

I would rather not change it in this PR, since it affects `%` and `WxH` chains that predate the side values. Should I open a separate issue for it?

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (Mikachu2333) [reply src/cli/pipeline.rs:509]: Yes, and fix it in another PR.

Comment (vlad-salone) [reply src/cli/pipeline.rs:509]: Filed as #379 with the three cases above.

One thing noted there for whoever picks it up: the intermediate resizes cannot simply be skipped in favour of the last target, because quantization and alpha premultiply can sit between two `--resize` values and do observe the intermediate size.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/cli/preprocessors/mod.rs (hunk 2 above)
Comment (Mikachu2333) [reply src/cli/preprocessors/mod.rs:58]: Add this to docs.

Comment (vlad-salone) [reply src/cli/preprocessors/mod.rs:58]: Documented in `294536e`, in the long help of both flags and in the README.

```
      --no-upscale
          Disable upscaling when resizing. [aliases: --reduce-only]

          This is useful when you don't want to increase the size of the image when it is smaller than the specified size.
          Images that are already smaller than the specified size are left untouched, so this is the reduce only mode.
          Passing it together with --no-downscale leaves no direction to resize in, so every image keeps its original size.
          It is recommended to use this option with --resize
```

The README carries the same note next to the resize examples, along with an `--enlarge-only` example that was missing.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Mikachu2333 — DISMISSED** [review]
CHANGELOG is requested.

**Comment by vlad-salone** [comment]
Added in `5aa6bcf`, under a new `Unreleased` heading since 0.12.4 is already out:

```md
# Unreleased

### Features

- add `100l` and `100s` resize values that anchor on the longest or shortest side of each image ([#377](https://github.com/SalOne22/rimage/pull/377))
- add `--reduce-only` and `--enlarge-only` as aliases for `--no-upscale` and `--no-downscale` ([#377](https://github.com/SalOne22/rimage/pull/377))
```

I left the strict parsing out of it. It only applies to the new `l` and `s` values, so there is nothing there that changes for anyone on 0.12.4.

Two notes on the file itself, neither of which I touched here. Every link in it points at `SalOne22/rimage` while the remote is now `vlad-salone/rimage`, and the same is true of `repository` in `Cargo.toml` and the README badges. The redirect covers it, so nothing is broken. Existing entries also link commits rather than pull requests, but a commit link needs the hash to be on `main`, which it is not until this merges.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by vlad-salone** [comment]
I guess I went a little too far with agentic automation 😬 but it's cool that it can answer for me when I'm unavailable

**Comment by Mikachu2333** [comment]
I think it's fine. When I use AI Agent to help me coding, I always impose fairly detailed constraints, and the final results are quite good. Of course, I carefully review everything, require test cases for all critical changes, and consider edge cases (it has to be said that AI always finds all sorts of strange edge cases, which greatly saves me time writing tests and significantly broadens coverage).

And, I'd use AI re-review the code before upload with a vague ptompt like:
```
Find and try to fix all of the following issues, report the issues first and requset for confirmation before making any changes:

1. All pointer, out-of-bounds, and memory issues
2. Program logic flaws, poor design, or ill-considered architecture
3. Race conditions
4. Deadlocks
5. Permission issues
6. FFI safety
7. Outdated functions or standards
8. Comments that do not match the code
9. Redundant code
10. Incomplete error handling that could lead to serious consequences
11. Undefined behavior
12. Compatibility issues
13. Unreasonable GUI layout
```

This truly helps me a lot for typos, ill-considered and so on.

---

By the way, in my opinion, it's wise to disable auto-upload and github-modify premission for agent to avoid any `prompt injection attack`. For example, I could never approve AI Agent auto-reply to other's msg and only allow it to generate reply locally and send it manually...

**Review by Mikachu2333 — APPROVED** [review]
